src/ingest_validation_tools/upload.py

```diff
@@ -278,38 +255,58 @@ def _get_local_tsv_errors(self) -> Optional[Dict]:
-                at the next instance of whitespace (in this case, leading
-                (as x) to be on a following line)
-                Wrote validation_utils > print_path to try to address this
-                issue, but needed to deprioritize.
-                """
-                errors.update(dir_errors)
+        if self.multi_assay_data_paths:
+            for path, dataset_types in self.multi_assay_data_paths.items():
+                dir_errors = self._get_multi_assay_dir_errors(path, dataset_types)
+                if dir_errors:
+                    errors.update(dir_errors)
+        else:
+            for path, schema in self.effective_tsv_paths.items():
+                dir_errors = self._get_ref_errors("data", schema, path)
+                if dir_errors:
+                    """
+                    TODO: there's an issue here (and any other places setting a key
+                    that might be long) where the YAML dumper converts this to a
+                    complex key and adds a ? at the front and inserts a line break
+                    at the next instance of whitespace (in this case, leading
+                    (as x) to be on a following line)
+                    Wrote validation_utils > print_path to try to address this
+                    issue, but needed to deprioritize.
+                    """
+                    errors.update(dir_errors)
         return errors
 
-    def validation_routine(
-        self,
-        report_type: ReportType = ReportType.STR,
-        tsv_paths: Dict[str, SchemaVersion] = {},
-    ) -> dict:
-        errors: DefaultDict[str, dict] = defaultdict(dict)
-        tsvs_to_evaluate = tsv_paths if tsv_paths else self.effective_tsv_paths
-        for tsv_path, schema_version in tsvs_to_evaluate.items():
-            path_errors = self._validate(tsv_path, schema_version, report_type)
-            if not path_errors:
-                return {}
-            for key, value in path_errors.items():
-                if type(value) is dict:
-                    errors[key].update(value)
-                else:
-                    errors.update({key: value})
-        return dict(errors)
+    def _get_multi_assay_dir_errors(
+        self, path: str, dataset_types: Dict
+    ) -> Optional[Dict]:
+        data_path = self.directory_path / path
+        parent = dataset_types.get("parent")
+        # Validate against parent multi-assay type if data_path is in parent TSV
+        if parent:
+            return self._multi_assay_dir_check(parent[0], data_path, path)
+        # Validate against component structure otherwise
+        elif dataset_types.get("components"):
+            errors = {}
+            for component in dataset_types["components"]:
+                errors.update(self._multi_assay_dir_check(component, data_path, path))
+            return errors
+
+    def _multi_assay_dir_check(
+        self, schema: SchemaVersion, ref_path: Path, data_path: str
```

Comment:
I'm a bit confused on the invocation on line 291 and the declaration here, data_path -> ref_path and path -> data_path

Reply:
Oh you're right that's terrible.

Reply:
Made a fix, let me know if that's better!